sensorhub-core/src/main/java/org/sensorhub/impl/comm/TCPServerCommProvider.java

```diff
@@ -0,0 +1,136 @@
+
+ Copyright (C) 2023 Botts Innovative Research Inc. All Rights Reserved.
+ ******************************* END LICENSE BLOCK ***************************/
+
+package org.sensorhub.impl.comm;
+
+import org.sensorhub.api.comm.ICommProvider;
+import org.sensorhub.api.common.SensorHubException;
+import org.sensorhub.impl.module.AbstractModule;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.net.ssl.SSLServerSocket;
+import javax.net.ssl.SSLServerSocketFactory;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.ServerSocket;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Consumer;
+
+/**
+ * Communication provider for TCP/IP server connections.
+ *
+ * @author Michael Elmore
+ * @since September 2023
+ */
+public class TCPServerCommProvider extends AbstractModule<TCPServerCommProviderConfig> implements ICommProvider<TCPServerCommProviderConfig> {
+    static final Logger log = LoggerFactory.getLogger(TCPServerCommProvider.class.getSimpleName());
+    private final Set<Consumer<ConnectionEventArgs>> eventHandlers = new HashSet<>();
+
+    ServerSocket serverSocket;
+    Thread serverThread;
+
+    public TCPServerCommProvider() {
+    }
+
+    @Override
+    protected void doStart() throws SensorHubException {
+        var config = this.config.protocol;
+
+        //Start a thread to listen on the configured port
+        serverThread = new Thread(() -> {
+            try {
+                if (config.enableTLS) {
+                    var sslServerSocketFactory = (SSLServerSocketFactory) SSLServerSocketFactory.getDefault();
+                    try (var sslServerSocket = (SSLServerSocket) sslServerSocketFactory.createServerSocket(config.localPort)) {
+                        while (true) {
+                            var socket = sslServerSocket.accept();
+                            broadcast(socket.getInputStream(), socket.getOutputStream());
+                        }
+                    }
+                } else {
+                    serverSocket = new ServerSocket(config.localPort);
+                    while (true) {
+                        var socket = serverSocket.accept();
+                        broadcast(socket.getInputStream(), socket.getOutputStream());
+                    }
+                }
+            } catch (Exception e) {
+                log.error("TCPServerCommProvider error: " + e.getMessage());
+            }
+        });
+        serverThread.start();
+    }
+
+    @Override
+    protected void doStop() throws SensorHubException {
+        try {
+            if (serverSocket != null) {
+                serverSocket.close();
+            }
+            if (serverThread != null) {
+                serverThread.interrupt();
+            }
+        } catch (IOException e) {
+            log.error("TCPServerCommProvider error: " + e.getMessage());
```

Comment:
You could throw a new SensorHubException here that is caught by the calling function or propagated